ce/api/geo.py

```diff
@@ -104,55 +100,25 @@ def cache_clear(self):
             self.hits = 0
             self.misses = 0
 
-def pointInPoly(x, y, poly):
-    if x is ma.masked or y is ma.masked:
-        return ma.masked
-
-    cell_center = Point(x, y)
-    if poly.contains(cell_center):
-        return not ma.masked
-    else:
-        return ma.masked
-
-pointsInPoly = np.vectorize(pointInPoly)
-
 @memoize_mask
-def wktToMask(nc, wkt):
+def wktToMask(filename, wkt, variable):
     poly = loads(wkt)
-    return polygonToMask(nc, poly)
+    return polygonToMask(filename, poly, variable)
 
-def polygonToMask(nc, poly):
+def polygonToMask(filename, poly, variable):
 
-    nclats = nc.variables['lat'][:]
+    nc = Dataset(filename, 'r')
     nclons = nc.variables['lon'][:]
     if np.any(nclons > 180):
-        nclons -= 180
-
-    lons, lats = np.meshgrid(nclons, nclats)
-    lons = ma.masked_array(lons)
-    lats = ma.masked_array(lats, mask=lons.mask)
-    # The mask is now shared between both arrays, so you can mask one and the
-    # other will be masked as well
-    assert lons.sharedmask == True
-
-    # Gather grid properties
-    min_width = np.min(np.diff(nclons))
-    min_height = np.min(np.diff(nclons))
-    min_cell_area = min_width * min_height
-
-    if poly.area < min_cell_area * 2:
-        # Buffer the polygon by 1/2 the diagonal distance
-        # FIXME: use something better than min_cell_area * 2.
-        # Maybe buffer based on function of selected size vs grid cell size to avoid hard cutoff
-        dist = math.sqrt(math.pow(min_width,2)+math.pow(min_height,2))/2
-        poly = poly.buffer(dist)
-
-    # Calculate the polygon extent
-    minx, miny, maxx, maxy = poly.bounds
+        poly = translate(poly, xoff=180)
+    nc.close()
 
-    lons = ma.masked_where((lons < minx) | (lons > maxx), lons, copy=False)
-    lats = ma.masked_where((lats < miny) | (lats > maxy), lats, copy=False)
+    dst_name = 'NETCDF:"{}":{}'.format(filename, variable)
+    raster = rasterio.open(dst_name, 'r', driver='NetCDF')
 
-    polygon_mask = pointsInPoly(lons, lats, poly)
+    if raster.affine == rasterio.Affine.identity():
+        raise Exception("Unable to determine projection parameters for GDAL "
+                        "dataset {}".format(dst_name))
 
-    return polygon_mask.mask
+    mask = rasterize((poly,), out_shape=raster.shape, transform=raster.affine, all_touched=True)
```

Comment:
Yeah, pretty nice. It still doesn't yet give us the RAT feature of weighting grid cells by percentage of each cell included, but I think that we can get that eventually (and our old code didn't give us that either).